File: lib/avro/avro_engine.dart
/// Avro-style Banglish to Bengali engine.
/// Longest-match patterns plus prefix/suffix context rules.
class AvroEngine {
  static const vowels = 'aeiou';
  static const consonants = 'bcdfghjklmnpqrstvwxyz';

  String parse(String input) {
    if (input.isEmpty) return '';
    final out = StringBuffer();
    final buf = StringBuffer();
    void flush() {
      if (buf.isEmpty) return;
      out.write(parseWord(buf.toString()));
      buf.clear();
    }

    for (var i = 0; i < input.length; i++) {
      final ch = input[i];
      if (_isLatinLetter(ch) || ch == '`') {
        buf.write(ch);
      } else {
        flush();
        out.write(ch);
      }
    }
    flush();
    return out.toString();
  }

  String parseWord(String raw) {
    if (raw.isEmpty) return '';
    if (raw.startsWith('`')) {
      return raw.substring(1);
    }
    final word = _fixCase(raw);
    final exception = _exceptions[word];
    if (exception != null) return exception;

    final out = StringBuffer();
    var i = 0;
    while (i < word.length) {
      final hit = _match(word, i);
      out.write(hit.replace);
      i += hit.consumed;
    }
    return out.toString();
  }

  _Hit _match(String word, int i) {
    for (final p in _patterns) {
      if (i + p.find.length > word.length) continue;
      if (word.substring(i, i + p.find.length) != p.find) continue;
      if (p.rules.isEmpty) {
        return _Hit(p.find.length, p.replace);
      }
      for (final rule in p.rules) {
        if (rule.matches.every((m) => _ok(m, word, i, i + p.find.length))) {
          return _Hit(p.find.length, rule.replace);
        }
      }
      return _Hit(p.find.length, p.replace);
    }
    return _Hit(1, word[i]);
  }

  bool _ok(_Match m, String word, int start, int end) {
    if (m.prefix) {
      if (start == 0) return false;
      return _check(m, word, start - 1);
    }
    if (end >= word.length) return false;
    return _check(m, word, end);
  }

  bool _check(_Match m, String word, int index) {
    if (index < 0 || index >= word.length) return false;
    final ch = word[index].toLowerCase();
    return switch (m.kind) {
      _Kind.vowel => AvroEngine.vowels.contains(ch),
      _Kind.consonant => AvroEngine.consonants.contains(ch),
    };
  }

  String _fixCase(String text) {
    final b = StringBuffer();
    for (final ch in text.split('')) {
      if ('OITDNSRZ'.contains(ch)) {
        b.write(ch);
      } else {
        b.write(ch.toLowerCase());
      }
    }
    return b.toString();
  }

  bool _isLatinLetter(String ch) {
    final c = ch.codeUnitAt(0);
    return (c >= 65 && c <= 90) || (c >= 97 && c <= 122);
  }
}

class AvroComposer {
  AvroComposer({AvroEngine? engine}) : engine = engine ?? AvroEngine();

  final AvroEngine engine;
  String buffer = '';

  String get preview => engine.parse(buffer);

  void type(String ch) => buffer += ch;

  void backspace() {
    if (buffer.isEmpty) return;
    buffer = buffer.substring(0, buffer.length - 1);
  }

  String commitWord() {
    final text = preview;
    buffer = '';
    return text;
  }
}

class _Hit {
  const _Hit(this.consumed, this.replace);
  final int consumed;
  final String replace;
}

enum _Kind { vowel, consonant }

class _Match {
  const _Match.prefix(this.kind) : prefix = true;
  const _Match.suffix(this.kind) : prefix = false;
  final bool prefix;
  final _Kind kind;
}

class _Rule {
  const _Rule(this.replace, this.matches);
  final String replace;
  final List<_Match> matches;
}

class _Pat {
  const _Pat(this.find, this.replace, [this.rules = const []]);
  final String find;
  final String replace;
  final List<_Rule> rules;
}

const _pv = _Kind.vowel;
const _pc = _Kind.consonant;

const _patterns = <_Pat>[
  _Pat('ksh', 'ক্ষ'),
  _Pat('cch', 'ছ'),
  _Pat('chh', 'ছ'),
  _Pat('ngo', 'ঙ্গ'),
  _Pat('nno', 'ন্য'),
  _Pat('nyo', 'ন্য'),
  _Pat('kh', 'খ'),
  _Pat('gh', 'ঘ'),
  _Pat('ch', 'চ'),
  _Pat('jh', 'ঝ'),
  _Pat('Th', 'ঠ'),
  _Pat('Dh', 'ঢ'),
  _Pat('th', 'থ'),
  _Pat('dh', 'ধ'),
  _Pat('ph', 'ফ'),
  _Pat('bh', 'ভ'),
  _Pat('sh', 'শ'),
  _Pat('Ng', 'ঙ'),
  _Pat('ng', 'ং'),
  _Pat('rh', 'ড়'),
  _Pat('Rh', 'ঢ়'),
  _Pat('aa', 'আ'),
  _Pat('ee', 'ঈ'),
  _Pat('oo', 'ঊ'),
  _Pat('oi', 'ঐ', [_Rule('ৈ', [_Match.prefix(_pc)])]),
  _Pat('ou', 'ঔ', [_Rule('ৌ', [_Match.prefix(_pc)])]),
  _Pat('a', 'আ', [_Rule('া', [_Match.prefix(_pc)])]),
  _Pat('i', 'ই', [_Rule('ি', [_Match.prefix(_pc)])]),
  _Pat('I', 'ঈ', [_Rule('ী', [_Match.prefix(_pc)])]),
  _Pat('u', 'উ', [_Rule('ু', [_Match.prefix(_pc)])]),
  _Pat('e', 'এ', [_Rule('ে', [_Match.prefix(_pc)])]),
  _Pat('o', 'ও', [
    _Rule('', [_Match.prefix(_pc), _Match.suffix(_pc)]),
    _Rule('ো', [_Match.prefix(_pc)]),
  ]),
  _Pat('O', 'ও', [_Rule('ো', [_Match.prefix(_pc)])]),
  _Pat('y', 'য', [_Rule('য়', [_Match.prefix(_pv)])]),
  _Pat('r', 'র', [_Rule('্র', [_Match.prefix(_pc)])]),
  _Pat('k', 'ক'),
  _Pat('g', 'গ'),
  _Pat('c', 'চ'),
  _Pat('j', 'জ'),
  _Pat('T', 'ট'),
  _Pat('D', 'ড'),
  _Pat('t', 'ত'),
  _Pat('d', 'দ'),
  _Pat('N', 'ণ'),
  _Pat('n', 'ন'),
  _Pat('p', 'প'),
  _Pat('f', 'ফ'),
  _Pat('b', 'ব'),
  _Pat('m', 'ম'),
  _Pat('l', 'ল'),
  _Pat('S', 'ষ'),
  _Pat('s', 'স'),
  _Pat('h', 'হ'),
  _Pat('z', 'য'),
  _Pat('w', 'ও'),
  _Pat('v', 'ভ'),
  _Pat('x', 'ক্স'),
  _Pat('q', 'ক'),
];

const _exceptions = <String, String>{
  'acho': 'আছো',
  'achho': 'আছো',
  'accho': 'আছো',
  'dhonnobad': 'ধন্যবাদ',
  'dhonnyobad': 'ধন্যবাদ',
};

File: android/app/src/main/kotlin/com/avrokeyboard/app/ime/avro/AvroEngine.kt
package com.avrokeyboard.app.ime.avro

/** Keep in sync with lib/avro/avro_engine.dart */
class AvroEngine {
    fun parse(input: String): String {
        if (input.isEmpty()) return ""
        val out = StringBuilder()
        val buf = StringBuilder()
        fun flush() {
            if (buf.isEmpty()) return
            out.append(parseWord(buf.toString()))
            buf.clear()
        }
        for (ch in input) {
            if ((ch.isLetter() && ch.code < 128) || ch == '`') buf.append(ch) else {
                flush()
                out.append(ch)
            }
        }
        flush()
        return out.toString()
    }

    fun parseWord(raw: String): String {
        if (raw.isEmpty()) return ""
        if (raw.startsWith("`")) return raw.substring(1)
        val word = fixCase(raw)
        exceptions[word]?.let { return it }
        val out = StringBuilder()
        var i = 0
        while (i < word.length) {
            val hit = match(word, i)
            out.append(hit.replace)
            i += hit.consumed
        }
        return out.toString()
    }

    private fun match(word: String, i: Int): Hit {
        for (p in patterns) {
            if (i + p.find.length > word.length) continue
            if (word.substring(i, i + p.find.length) != p.find) continue
            if (p.rules.isEmpty()) return Hit(p.find.length, p.replace)
            for (rule in p.rules) {
                if (rule.matches.all { ok(it, word, i, i + p.find.length) }) {
                    return Hit(p.find.length, rule.replace)
                }
            }
            return Hit(p.find.length, p.replace)
        }
        return Hit(1, word[i].toString())
    }

    private fun ok(m: Match, word: String, start: Int, end: Int): Boolean =
        if (m.prefix) {
            if (start == 0) false else check(m, word, start - 1)
        } else {
            if (end >= word.length) false else check(m, word, end)
        }

    private fun check(m: Match, word: String, index: Int): Boolean {
        if (index !in word.indices) return false
        val ch = word[index].lowercaseChar()
        return when (m.kind) {
            Kind.VOWEL -> ch in vowels
            Kind.CONSONANT -> ch in consonants
        }
    }

    private fun fixCase(text: String): String = buildString {
        for (ch in text) append(if (ch in "OITDNSRZ") ch else ch.lowercaseChar())
    }

    private data class Hit(val consumed: Int, val replace: String)
    private enum class Kind { VOWEL, CONSONANT }
    private data class Match(val prefix: Boolean, val kind: Kind)
    private data class Rule(val replace: String, val matches: List<Match>)
    private data class Pat(val find: String, val replace: String, val rules: List<Rule> = emptyList())

    companion object {
        private const val vowels = "aeiou"
        private const val consonants = "bcdfghjklmnpqrstvwxyz"
        private fun pV() = Match(true, Kind.VOWEL)
        private fun pC() = Match(true, Kind.CONSONANT)
        private fun sC() = Match(false, Kind.CONSONANT)
        private val patterns = listOf(
            Pat("ksh", "ক্ষ"), Pat("ggy", "জ্ঞ"), Pat("GG", "জ্ঞ"),
            Pat("cch", "ছ"), Pat("chh", "ছ"),
            Pat("ngo", "ঙ্গ"), Pat("nno", "ন্য"), Pat("nyo", "ন্য"),
            Pat("sw", "স্ব"),
            Pat("kh", "খ"), Pat("gh", "ঘ"), Pat("ch", "চ"), Pat("jh", "ঝ"),
            Pat("Th", "ঠ"), Pat("Dh", "ঢ"), Pat("th", "থ"), Pat("dh", "ধ"),
            Pat("ph", "ফ"), Pat("bh", "ভ"), Pat("sh", "শ"),
            Pat("Ng", "ঙ"), Pat("ng", "ং"), Pat("rh", "ড়"), Pat("Rh", "ঢ়"),
            Pat("aa", "আ"), Pat("ee", "ঈ"), Pat("oo", "ঊ"),
            Pat("oi", "ঐ", listOf(Rule("ৈ", listOf(pC())))),
            Pat("ou", "ঔ", listOf(Rule("ৌ", listOf(pC())))),
            Pat("a", "আ", listOf(Rule("া", listOf(pC())))),
            Pat("i", "ই", listOf(Rule("ি", listOf(pC())))),
            Pat("I", "ঈ", listOf(Rule("ী", listOf(pC())))),
            Pat("u", "উ", listOf(Rule("ু", listOf(pC())))),
            Pat("e", "এ", listOf(Rule("ে", listOf(pC())))),
            Pat("o", "ও", listOf(Rule("", listOf(pC(), sC())), Rule("ো", listOf(pC())))),
            Pat("O", "ও", listOf(Rule("ো", listOf(pC())))),
            Pat("y", "য", listOf(Rule("য়", listOf(pV())))),
            Pat("r", "র", listOf(Rule("্র", listOf(pC())))),
            Pat("k", "ক"), Pat("g", "গ"), Pat("c", "চ"), Pat("j", "জ"),
            Pat("T", "ট"), Pat("D", "ড"), Pat("t", "ত"), Pat("d", "দ"),
            Pat("N", "ণ"), Pat("n", "ন"), Pat("p", "প"), Pat("f", "ফ"),
            Pat("b", "ব"), Pat("m", "ম"), Pat("l", "ল"),
            Pat("S", "ষ"), Pat("s", "স"), Pat("h", "হ"),
            Pat("z", "য"), Pat("w", "ও"), Pat("v", "ভ"),
            Pat("x", "ক্স"), Pat("q", "ক"),
        )
        private val exceptions = mapOf(
            "acho" to "আছো",
            "achho" to "আছো",
            "accho" to "আছো",
            "dhonnobad" to "ধন্যবাদ",
            "dhonnyobad" to "ধন্যবাদ",
        )
    }
}

class AvroComposer(private val engine: AvroEngine = AvroEngine()) {
    var buffer: String = ""
        private set
    val preview: String get() = engine.parse(buffer)
    fun type(ch: String) { buffer += ch }
    fun backspace() { if (buffer.isNotEmpty()) buffer = buffer.dropLast(1) }
    fun commitWord(): String { val t = preview; buffer = ""; return t }
    fun clear() { buffer = "" }
    val isComposing: Boolean get() = buffer.isNotEmpty()
}
